Merge dev into master

dmitry-zhbankov opened the pull request.

dmitry-zhbankov committed:
Add code

dmitry-zhbankov committed:
Add partial application implementation

dmitry-zhbankov committed:
Add functions

dmitry-zhbankov committed:
Add inheritance puzzle
Add tests

dmitry-zhbankov committed:
Add lazy thunk implementation

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+
+    context.curry = function (f) {
+        return function currify() {
+            const args = Array.prototype.slice.call(arguments);
```
Why don't you use Rest Parameters everywhere?

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+            accum = callback(accum, array[i], i, initialValue);
+        }
+        return accum;
+    };
```
Please, answer, why you have to use such methods in the following way:
linFold(array, (res, curr) => ({
    ...res,
    [sum] : res.sum + curr
}), { sum: 0 });

Instead of:
linFold(array, (res, curr) => {
    res.sum += curr;
    return res;
}, { sum: 0 });

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+let funcLib = {};
+
+arrLib = (function (context) {
+    context.partial = function (g, ...args1) {
```
Is there any JavaScript built-in alternative?

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+        return new ShapeStore();
+    };
+
+    class Shape {
```
Why don't you use functions you implemented above?

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+        }
+    };
+
+    context.multiply = function (...args) {
```
Why don't you use Linear fold solution?

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+
+    context.memo = function (f) {
+        let mem = {};
+        return function (n) {
```
It would be more readable if there is only one return:

if (!(n in mem)) {
	mem[n] = f(n);
}
return mem[n];

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+        let mem = {};
+        return function (n) {
+            if (n in mem) {
+                return mem[n];
```
Doesn't work for several parameters and if any parameter is Object

valerydolyuk commented on func-puzzles.js:
```diff
@@ -0,0 +1,177 @@
+    };
+
+    context.lazy = function (f) {
+        return function (...args) {
```
What you forgot is to store (memoize) the result.
Here u can find an example and also the problem of lazy functions: https://www.codementor.io/@agustinchiappeberrini/lazy-evaluation-and-javascript-a5m7g8gs3

dmitry-zhbankov committed:
Fix some issues:
Use rest parameters
Implement memoization for objects and multiple arguments
Reuse existing functions
